glib/src/bridged_logging.rs

```diff
@@ -265,6 +265,74 @@ pub fn rust_log_handler(domain: Option<&str>, level: glib_log::LogLevel, message
     rs_log::log!(target: domain.unwrap_or("<null>"), level, "{}", message);
 }
 
+// rustdoc-stripper-ignore-next
+/// Provides a glib log writer which routes all structured logging messages to the
+/// [`log crate`](https://crates.io/crates/log).
+///
+/// In order to use this function, `glib` must be built with the `log` feature
+/// enabled.
+///
+/// Use this function if you want to use the log crate as the main logging
+/// output in your application, and want to route all structured logging happening in
+/// glib to the log crate. If you want the opposite, use [`GlibLogger`](struct.GlibLogger.html).
+///
+/// NOTE: This should never be used when [`GlibLogger`](struct.GlibLogger.html) is
+/// registered as a logger, otherwise a stack overflow will occur.
+///
+/// ```no_run
+/// glib::log_set_writer_func(glib::rust_log_writer);
+/// ```
+pub fn rust_log_writer(log_level: glib_log::LogLevel, fields: &[LogField<'_>]) -> LogWriterOutput {
+    let lvl = match log_level {
+        glib_log::LogLevel::Error | glib_log::LogLevel::Critical => rs_log::Level::Error,
+        glib_log::LogLevel::Warning => rs_log::Level::Warn,
+        glib_log::LogLevel::Info | glib_log::LogLevel::Message => rs_log::Level::Info,
+        glib_log::LogLevel::Debug => rs_log::Level::Debug,
+    };
+
+    if lvl > rs_log::STATIC_MAX_LEVEL {
+        return LogWriterOutput::Handled;
+    }
+
+    let mut domain = None::<&str>;
+    let mut message = None::<&str>;
+    let mut file = None::<&str>;
+    let mut line = None::<u32>;
+    let mut func = None::<&str>;
+
+    for field in fields {
+        let Some(value) = field.value_str() else {
+            continue;
+        };
+
+        match field.key() {
+            "GLIB_DOMAIN" => domain = Some(value),
+            "MESSAGE" => message = Some(value),
+            "CODE_FILE" => file = Some(value),
+            "CODE_LINE" => line = value.parse().ok(),
+            "CODE_FUNC" => func = Some(value),
+            _ => continue,
```

Comment:
`log` also has key-value support. Could all the other keys here be mapped to that?

Reply:
Structured logging in `log` is disabled by default so I'd want to gate it behind a `log-kv` or `log-structured` feature.

The actual implementation is pretty trivial. We'd just implement the `log::Source` trait for `glib::LogField` and add the whole array of fields to the `RecordBuilder`.